### DISCUSSION TITLE
tasmota-ir and tasmota-ircustom confusion

### DISCUSSION THREAD

**Discussion by benzino77** [arendst/Tasmota#10790] [Q&A] [answered]
Hi there,

`platformio_tasmota_env.ini` file contains:
```
[env:tasmota-ir]
build_flags             = ${common.build_flags} ${irremoteesp_full.build_flags} -DFIRMWARE_IR
lib_extra_dirs          = lib/lib_basic

[env:tasmota-ircustom]
build_flags             = ${common.build_flags} ${irremoteesp_full.build_flags} -DFIRMWARE_IR_CUSTOM
```

but there are no `FIRMWARE_IR` and `FIRMWARE_IR_CUSTOM` sections in `tasmota/tasmota_configurations.h`. 

Additionally both envs have `build_flags` set to  `${irremoteesp_full.build_flags}` which includes `-DUSE_IR_REMOTE_FULL`. 

Maybe I miss something but it will never let compile `tasmota/xdrv_05_irremote.ino`.

**Comment by barbudor** [comment]
I can't either build `tasmota-ir` 
It fails on missing `LinkedList.h`

**Reply by benzino77** [reply]
tasmota-ir fails because it overwrite `lib_extra_dirs` and include only `lib/lib_basic`
`LinkedList.h` is located in `lib/lib_i2c` so it should be (at least):
```
lib_extra_dirs          = lib/lib_basic, lib/lib_i2c
```

**Reply by barbudor** [reply]
I know why LinkedList is missing
But I don't understand how tasmota-ir.bin is still generated daily but we can't build it ourselves

**Reply by benzino77** [reply]
We have to wait for @Jason2866 - I think he is the "Mastablasta" of *.ini files magic.

**Reply by barbudor** [reply]
@benzino77 I've looked a little bit and the only difference I see is that `tasmota-ir` is build with a lot of non-IR-related  features removed while `tasmota-ircustom` is build on top of `tasmota`.
But because both are using `${irremoteesp_full.build_flags}` they are totally equivalent for IR features

My own custom build for my IR Blaster is based on `[env:tasmota-ircustom]`

I would suggest to simply forget about `tasmota-ir`

**Reply by barbudor** [reply]
@benzino77 I found why I couldn't compile `tasmota-ir`
It was because there was features added in my `user_config_override` while I was believing it to be empty.
In that case it is normal not being able to compile anything else than `tasmota`, or in that case `tasmota-ircustom`.
As specified in the Compiling-your-own-build page, you cannot customize any of the variants (except for basic Wifi/IP/MQTT parameters and setoptions) by adding features.
So in conclusion, no problems, everything is as expected and documented.

Removing the 2 defines in my `user_config_override.h` solved the problem.

**Reply by benzino77** [reply]
Well, it is not that simple for me. 
First there is no such thing like `tasmota-ir` and `tasmota-ircustom` because, as you noticed, they are, in actual configuration,  the same.
Second, in TasmoCompiler there is possibility to choose between "light" IR support and "full" IR support. I have to know whether to get rid of that option and just leave "IR support" without "light" and "full" support - mainly that was a background of my question.

**Reply by sfromis** [reply]
A key difference between tasmota-ir and tasmota-ircusom is that the former build environment does not allow adding features, the purpose of the latter is specifically to allow you to add features.

**Reply by barbudor** [reply]
@benzino77 
Sorry I didn't recognized you as TasmoCompiler owner at first ;)

I would expect that TasmoCompiler just like any custom-compile environment is always building on top of default `tasmota`, so the pre-defined variants doesn't make much senses for you.
Any box checked should enable features on top of Tasmota won't they ?
So you still have the ability to offer USE_REMOTE and USE_REMOTE_FULL independently 
But as you said, may be worth to only have 1 single option to be aligned with tasmota pre-defined variants.

**Reply by benzino77** [reply]
TC is building on top of `tasmota` as you said, and it is using its' own platformio env. I'm working on new version of TC and find this confusing. From my point of view leaving only one option to choose - just "IR support" is much more clearer for the end user than two options "basic" and "full".

**Comment by Jason2866** [comment]
LinkedList is only needed for rules "IF" enhancement.

This entry in Platfrmio.ini enables all IR Protocols 
```
[irremoteesp_full]
build_flags                 = -DUSE_IR_REMOTE_FULL
                              -U_IR_ENABLE_DEFAULT_
                              -DDECODE_PRONTO=false -DSEND_PRONTO=false
```

**Comment by Jason2866** [comment]
So for your use case @benzino77 just add this build flags to the base variant Tasmota and you have enabled all IR-Protocols

**Reply by benzino77** [reply]
@Jason2866 so there is no "light" and "full" version of IR. It is now just "IR support"?

**Comment by Jason2866** [comment, marked as answer]
Not possible to have only one version. IR full variant is to big to include in standard builds.
In standard build the most usual IR protocols are enabled.
So, IR basic protocols are included by default. 
Variant Tasmota-IR includes all IR protocols and can not be customized
Variant Tasmota-IR custom allow all modifications (like variant Tasmota) and includes all IR protocols
Default enabled IR protocols
```
-D_IR_ENABLE_DEFAULT_=false
                              -DDECODE_HASH=true -DDECODE_NEC=true -DSEND_NEC=true
                              -DDECODE_RC5=true -DSEND_RC5=true -DDECODE_RC6=true -DSEND_RC6=true
; new mechanism to set the IRremoteESP8266 supported protocols: none except HASH, NEC, RC5, RC6
```
